### DISCUSSION TITLE
Does ClearProvidersExceptFunctionProviders work as expected ?

### DISCUSSION THREAD

**Discussion by fgheysels** [arcus-azure/arcus.observability#189] [Q&A] [answered]
Since I'm seeing some strange behavior when using Arcus Observability in my Azure Function (custom events are written as an event and as a Warning  Trace, I decided to have a look at the `ClearProvidersExceptFunctionProviders` method and I saw this:

```csharp
foreach (ServiceDescriptor serviceDescriptor in loggingBuilder.Services)
{
    if (serviceDescriptor.ServiceType == typeof(ILoggerProvider))
    {
        if (serviceDescriptor.ImplementationType.FullName != "Microsoft.Azure.WebJobs.Script.Diagnostics.HostFileLoggerProvider"
                        && serviceDescriptor.ImplementationType.FullName != "Microsoft.Azure.WebJobs.Script.Diagnostics.FunctionFileLoggerProvider")
        {
            loggingBuilder.Services.Remove(serviceDescriptor);
            break;
        }
    }
}
```
[link](https://github.com/arcus-azure/arcus.observability/blob/9ff3ec96d8e0a3a6244ddbe2d1d04bb7cf42fe4f/src/Arcus.Observability.Telemetry.AzureFunctions/Extensions/ILoggerBuilderExtensions.cs#L31)
I'm wondering if it is correct to break out of the loop as soon as a serviceDescriptor is removed.   I mean, when we encounter an ILoggingProvider, we remove it if it is not one of those 2 specified.   However, can't we have multiple ILoggingProviders that should be removed ?  Why are we leaving the loop after the first ILoggerProvider that is encountered and that can be deleted ?

I was in doubt whether to open an issue or a discussion for this, but choose for discussion.

**Comment by stijnmoreels** [comment]
To be honest, I don't know anymore what the reason was why we added this logging extension. It was a problem that @tomkerkhove came across in an Azure Functions environment and solved it with this.

**Reply by tomkerkhove** [reply]
It was added because Azure Functions default logging providers where breaking our dimensions.

Everything worked fine, but it never reported the custom dimensions. We can revise this after our refactoring.

**Reply by fgheysels** [reply]
See also this: https://github.com/arcus-azure/arcus.observability/issues/336

**Comment by stijnmoreels** [comment, marked as answer]
Discussed offline: we will deprecate this method and eventually remove this Azure Functions-specific package entirely. #346
